pysd/py_backend/builder.py

```diff
@@ -128,53 +150,125 @@ def build_element(element, subscript_dict):
 
     """
     if element['kind'] == 'constant':
-        cache_type = "@cache('run')"
-    elif element['kind'] in ['setup', 'stateful']:  # setups only get called once, caching is wasted
-        cache_type = ''
-    elif element['kind'] == 'component':
-        cache_type = "@cache('step')"
-    elif element['kind'] == 'stateful':
+        cache_type = "@cache.run"
+    elif element['kind'] in ['component', 'component_ext_data']:
+        cache_type = "@cache.step"
+    elif element['kind'] == 'lookup':
+        # lookups may be called with different values in a round
         cache_type = ''
-    elif element['kind'] == 'lookup':  # lookups may be called with different values in a round
+    elif element['kind'] in ['setup', 'stateful',
+                             'external', 'external_add']:
+        # setups only get called once, caching is wasted
         cache_type = ''
     else:
         raise AttributeError("Bad value for 'kind'")
 
-    if len(element['py_expr']) > 1:
-        contents = 'return utils.xrmerge([%(das)s,])' % {'das': ',\n'.join(element['py_expr'])}
+    # check the elements with ADD in their name
+    # as these wones are directly added to the
+    # objcet via .add method
+    py_expr_no_ADD = ["ADD" not in py_expr for py_expr in element['py_expr']]
+
+    if sum(py_expr_no_ADD) > 1:
```

Comment:
Can you explain this to me a bit more? What is special about ADD? does it correspond to a summation, and is there likely to be a MULTIPLY that we need to treat in the same way? Or does it refer to something which was added on extra to the model?

Reply:
When an object is defined from external data in several lines only one object is created. Then the following vensim blocks are added to this object using `.add `method. This makes simpler the code, and also avoids merging the objects each time step as they are merged when they are initialized. You can check the final result in _test_get_lookups_data_3d_xls.py_ translated file from test-models repo.
Just to identify better these objects, as their python identifier is unique and no capital letters are used and their I added "ADD" to the end of their py_expr